File: apps/mobile/src/features/core/native-core-adapters/conversation.ts
import type { Conversation } from '@glimpse/shared';

import { nativeCoreBridgeHelpers } from '../native-core-bridge';
import type { BridgeCoreClient } from '../types';

import type { NativeAdapterDeps } from './common';

const { toConversationPatch } = nativeCoreBridgeHelpers;

export function createConversationAdapter(
  deps: NativeAdapterDeps,
): Pick<
  BridgeCoreClient,
  'createConversation' | 'listConversations' | 'updateConversation' | 'deleteConversation'
> {
  const { fallbackClient, runCoreAsync } = deps;

  return {
    async createConversation(conversation: Conversation): Promise<Conversation> {
      return runCoreAsync(
        (client) => client.createConversation(conversation),
        () => fallbackClient.createConversation(conversation),
      );
    },

    async listConversations(): Promise<Conversation[]> {
      return runCoreAsync(
        (client) => client.listConversations(),
        () => fallbackClient.listConversations(),
      );
    },

    async updateConversation(
      conversationId: string,
      patch: Partial<Conversation>,
    ): Promise<Conversation> {
      return runCoreAsync(
        (client) => client.updateConversation(conversationId, toConversationPatch(patch)),
        () => fallbackClient.updateConversation(conversationId, patch),
      );
    },

    async deleteConversation(
      conversationId: string,
      deletedAt: number,
    ): Promise<void> {
      await runCoreAsync(
        (client) => client.deleteConversation(conversationId, deletedAt),
        () => fallbackClient.deleteConversation(conversationId, deletedAt),
      );
    },
  };
}

File: packages/core-rust/src/application/conversation.rs
use crate::error::Result;
use crate::models::{Conversation, ConversationPatch};

use super::SharedCore;

impl SharedCore {
    pub fn create_conversation(&self, conversation: &Conversation) -> Result<Conversation> {
        self.client().create_conversation(conversation)
    }

    pub fn list_conversations(&self) -> Result<Vec<Conversation>> {
        self.client().list_conversations()
    }

    pub fn update_conversation(
        &self,
        conversation_id: &str,
        patch: &ConversationPatch,
    ) -> Result<Conversation> {
        self.client().update_conversation(conversation_id, patch)
    }

    pub fn delete_conversation(&self, conversation_id: &str, deleted_at: i64) -> Result<()> {
        self.client()
            .delete_conversation(conversation_id, deleted_at)
    }
}
